Note AI 요약기능 추가 및 리팩터링

monda95 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #
- 작업 요약: 어떤 작업을 했는지 간단하게 적어주세요.

## 📄 상세 내용
- ✨ AI 요약 기능 추가 (Gemini)
  - SDK google.genai로 호출하는 방식으로 구현
  - `generate_ai_summary_task.delay(note.id)` 방식으로 Celery 비동기 실행
  - Note 생성 시 AI 요약 자동 생성 후 `ai_summary` 필드 저장
  - 게시글 수정 시 `content` 변경 감지 → 자동 재요약 (요구사항 외 필요 기능이라 판단)
  - 추후 ‘요약 재생성 버튼’ 또는 조건부 트리거 방식으로 개선 예정
  - `StudyNoteService` 내부 `genai.Client` 인스턴스 재사용 구조 유지
  - `self_check_summary` 메서드 추가 → 1차 요약이 불완전할 경우 자동 보정 수행
  
- ⚙️ **환경 및 설정**
  - `.env` 기반 전역 설정(`GEMINI_API_KEY`, `GEMINI_MODEL_NAME`) 추가  
  - `.local.env` 환경 지원 및 테스트 시 Mock 객체로 대체 (CI 통과 보장)

- 🧩 **Serializer / View 개선**
  - DetailSerializer에 파일·이미지 URL 필드 추가
  - ListSerializer에서는 불필요한 필드(`group_uuid`) 제거

- 🧪 테스트 추가 및 수정
  - Gemini API 호출에 대한 Mock Test 추가
  - 수정 시 AI 재요약 태스크 실행 여부 검증 테스트 작성

- 💬 **코드 리팩터링**
  - Note API 주석 및 docstring 일괄 수정
  - 관련 로직 주석/로그 명확화


## 📸 스크린샷 (선택)

## 📝 기타 참고 사항
- SDK는 인증·요청 구성을 내부적으로 처리하며, `genai.Client` 재사용으로 코드 간결성과 안정성 등이 높다고하여 채택. 단점으로는, HTTP 호출 방식과 달리 내부로직이 블랙박스인 점.
- 모델명(`gemini-2.5-flash`)은 동일.

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

monda95 committed:
feat : AI 요약용 호출로직

- google genai를 활용한 노트 내용요약용 AI 호출로직 추가

- urls group_id -> _uuid로 수정

# Conflicts:
#	apps/studies/serializers/notes.py

# Conflicts:
#	apps/studies/services/notes.py

# Conflicts:
#	poetry.lock
#	pyproject.toml

monda95 committed:
fix : config 타입 수정

- genai 추가 이후 cofig를 python dict가 아닌 GenerateContentConfig 객체로 받아야해서 변경

monda95 committed:
feat : Note AI 요약 기능

- service/notes.py에 구현하고
 DetailView에서 호출

- “RESTful HTTP 요청을 직접 전송하지 않고, 구글의 공식 SDK인 google-genai를 사용하여 클라이언트를 초기화하고, 이후 동일 세션 내에서 재사용하는 방식으로 AI 요약 로직을 구현
(boto3 클라이언트와 같은 원리라고 이해하고 진행)

monda95 committed:
feat : Note AI 요약 셀프체크추가 + celery task로 호출 및 그 외 수정사항

- generate_ai_summary_task.delay(note.id) 로 태스크 큐에 위임 + 요구사항엔 없지만 게시글 수정 시 AI 요약 재생성(재생성 버튼 or 재생성 조건문추가 등으로 업그레이드 고려중)

- serializers/notes 필드 재조정 (첨부할 파일/이미지에 대한 필드{url} 등 추가)

- ListView에서 필드 재조정 list에서는 필요없는 필드라 판단되는 group_uuid 제거 등

- Gemini mock test 추가

- dotenv 사용프로젝트 기준 Gemini API전역설정추가

- self_check_summary 추가
1차 프롬프트가 제대로 발동되지 않았을 경우에 작동하여 다시 한번 답변을 깎아서 제대로 요구된  답을 내도록하는 이중장치역할

Meoyoug commented on apps/studies/views/notes.py:
```diff
@@ -79,13 +75,22 @@ def post(self, request: Request) -> Response:
+        except Exception as e:
+            logger.error(
+                "[NoteCreate] AI 요약 비동기 호출 실패: note_id=%s, error=%s", note.id, e, exc_info=True
+            )  # ex) [NoteCreate] AI 요약 비동기호출 실패 : "1", "{e}" 레이즈 방식이라 가볍고 모든 걸 str로 타입변환
```
이거 그냥 바로 로직 처리하는걸로 호출해야할 것 같아요
셀러리로 처리하면 사용자가 노트를 저장한 후에 바로 해당 노트 조회페이지로 이동할텐데
즉시 AI summary를 조회할 수 없을 것 같아요

Meoyoug commented on config/settings/base.py:
```diff
@@ -287,3 +287,7 @@
 APPEND_SLASH = False
+
+# 외부 API 키
+GEMINI_API_KEY = os.getenv("GEMINI_API_KEY", "")
+GEMINI_MODEL_NAME = os.getenv("GEMINI_MODEL_NAME", "gemini-2.5-flash")
```
이거 환경변수는 저한테 디엠으로 전달부탁드립니다.

Meoyoug commented on apps/studies/permissions.py:
```diff
@@ -37,9 +37,20 @@ def has_permission(self, request: Request, view: APIView) -> bool:
+        # 노트 객체일 때
+        elif isinstance(obj, StudyNote):
+            return obj.study_group.group_members.filter(user=user).exists()
+
```
이거 굳이 추가안하고 check_object_permission호출할때 obj.study_group을 넘겨주면 되지않나요?

monda95 committed:
fix : permission 복구 및 CI 통과용 조건문 추가

- permission note 객체용을 추가 했던 것을 복구하고 객체를note 에서 note.study_group으로 다시 변경

monda95 committed:
fix : 비동기 -> 동기로 변경

- 비동기 task 삭제 후 동기로 변경

- 리팩터링 (필요없는 프리패치제거)

monda95 committed:
fix : view 리팩터링

- 5팀 응답규약 믹신으로 클래스 상속시켜서 간결화

- note author 검증이 돌아갈 때 객체를 그룹에서 찾고, 그룹 검증이 돌아갈 때 note 객체를 못 찾고 note.study_group만 넘겨서 해결하는 법을 도저히 못 찾겠어서 퍼미션별로 나눠서 해결

- 브랜치 최신화
